compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1218,9 +1218,13 @@ impl<'tcx> InferCtxt<'tcx> {
         );
 
         if self.tcx.sess.err_count() > self.err_count_on_creation {
-            return true; // errors reported since this infcx was made
+            // errors reported since this infcx was made
+            return Some(self.tcx.sess.delay_span_bug(
+                DUMMY_SP,
+                "`tcx.sess.error_count()` incorrectly returned non zero value",
+            ));
         }
-        self.tainted_by_errors.get().is_some()
+        self.tainted_by_errors.get()
```

Comment:
Could you swap these? It's cheaper to avoid `delay_span_bug` entirely, when `tainted_by_errors` is `Some`.